<a href="https://colab.research.google.com/github/Jojomaque/DataScience/blob/main/JojoDescribing_data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Describing data with summary statistics
---

This worksheet is tied to the Pandas Getting Started Tutorials, picking out particular tutorials to link them into a theme here.

We will focus on describing data.  This is the least risky in terms of bias and inaccurate conclusions as it should focus just on what data is presented to us.

Each exercise will ask you to work through on tutorial on the Getting Started page, to try the code from the tutorial here and to try a second, similar action.

---

The practice data from the tutorials comes from a dataset on Titanic passengers.


### Exercise 1 - open the Titanic dataset
---

The Titanic dataset is stored at this URL:
https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv

Read the dataset into a pandas dataframe that you will call **titanic**.

**Test output**:  
The shape of the dataframe will be (891, 12)

In [2]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv')
titanic = df
titanic.head(50)

PassengerId  Survived  Pclass  ...      Fare        Cabin  Embarked
0             1         0       3  ...    7.2500          NaN         S
1             2         1       1  ...   71.2833          C85         C
2             3         1       3  ...    7.9250          NaN         S
3             4         1       1  ...   53.1000         C123         S
4             5         0       3  ...    8.0500          NaN         S
5             6         0       3  ...    8.4583          NaN         Q
6             7         0       1  ...   51.8625          E46         S
7             8         0       3  ...   21.0750          NaN         S
8             9         1       3  ...   11.1333          NaN         S
9            10         1       2  ...   30.0708          NaN         C
10           11         1       3  ...   16.7000           G6         S
11           12         1       1  ...   26.5500         C103         S
12           13         0       3  ...    8.0500          NaN         S
13           14         0       3  ...   31.2750          NaN         S
14           15         0       3  ...    7.8542          NaN         S
15           16         1       2  ...   16.0000          NaN         S
16           17         0       3  ...   29.1250          NaN         Q
17           18         1       2  ...   13.0000          NaN         S
18           19         0       3  ...   18.0000          NaN         S
19           20         1       3  ...    7.2250          NaN         C
20           21         0       2  ...   26.0000          NaN         S
21           22         1       2  ...   13.0000          D56         S
22           23         1       3  ...    8.0292          NaN         Q
23           24         1       1  ...   35.5000           A6         S
24           25         0       3  ...   21.0750          NaN         S
25           26         1       3  ...   31.3875          NaN         S
26           27         0       3  ...    7.2250          NaN         C
27           28         0       1  ...  263.0000  C23 C25 C27         S
28           29         1       3  ...    7.8792          NaN         Q
29           30         0       3  ...    7.8958          NaN         S
30           31         0       1  ...   27.7208          NaN         C
31           32         1       1  ...  146.5208          B78         C
32           33         1       3  ...    7.7500          NaN         Q
33           34         0       2  ...   10.5000          NaN         S
34           35         0       1  ...   82.1708          NaN         C
35           36         0       1  ...   52.0000          NaN         S
36           37         1       3  ...    7.2292          NaN         C
37           38         0       3  ...    8.0500          NaN         S
38           39         0       3  ...   18.0000          NaN         S
39           40         1       3  ...   11.2417          NaN         C
40           41         0       3  ...    9.4750          NaN         S
41           42         0       2  ...   21.0000          NaN         S
42           43         0       3  ...    7.8958          NaN         C
43           44         1       2  ...   41.5792          NaN         C
44           45         1       3  ...    7.8792          NaN         Q
45           46         0       3  ...    8.0500          NaN         S
46           47         0       3  ...   15.5000          NaN         Q
47           48         1       3  ...    7.7500          NaN         Q
48           49         0       3  ...   21.6792          NaN         C
49           50         0       3  ...   17.8000          NaN         S

[50 rows x 12 columns]

### Exercise 2 - get summary information about the dataframe
---

Read through the tutorials:  
[What kind of data does pandas handle?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/01_table_oriented.html#)  
[How do I read and write tabular data?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/02_read_write.html)

Use panda functions to display the following:
1.  A technical summary of the data (info())
2.  A description of the numerical data (describe())
3.  Display the Series 'Age'

**Test output**:   
1.  The info should show that there are only 204 values in the Cabin series, out of 891 records.  
2.  The description should show 7 columns and a mean age of 29.699118
3.  The Age series should have values of type float64 and Length 891

In [ ]:
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv')
titanic = df
titanic.info()

In [ ]:
titanic.describe()

In [ ]:
titanic["Age"] #Looking at one column

### Exercise 3 - aggregating statistics
---

Read through the tutorial:  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

Use panda functions to display the following summary statistics from the titanic dataset:  

1.  The average (mean) age of passengers  
2.  The median age and fare  
3.  The mean fare
4.  The modal fare and gender

**Test output**:   
29.699118, Age 28.0000 Fare 14.4542, 32.2042079685746, Fare 8.05 Sex male 


In [ ]:
titanic["Age"].mean() #calculating the mean

In [ ]:
titanic[["Age", "Fare"]].median()

In [ ]:
titanic['Fare'].mean()

In [ ]:
titanic[["Fare", "Sex"]].mode()

### Exercise 4 - displaying other statistics
---

Take a look at the list of methods available for giving summary statistics [here](https://pandas.pydata.org/docs/user_guide/basics.html#basics-stats) 

Use panda functions, and your existing knowledge, to display the following summary statistics from the titanic dataset:

1.  The total number of passengers on the titanic
2.  The age of the youngest passenger
3.  The most expensive ticket price
4.  The range of ticket prices
5.  The number of passenges with cabins
6.  The code for the port where the highest number of passengers embarked
7.  The most populous gender
8.  The standard deviation for age and fare

**Tests**:  
891, 0.42, 512.3292, 512.3292, 204, S, male, Age 14.526497 Fare 49.693429

In [ ]:
titanic['PassengerId'].count() #The total number of passengers on the titanic

In [ ]:
titanic["Age"].min()

In [ ]:
titanic["Fare"].max()

In [ ]:
max_fare = titanic["Fare"].max()
min_fare = titanic ["Fare"].min()
fare_range = max_fare - min_fare
print(fare_range)

In [ ]:
titanic["Cabin"].value_counts().sum()

In [ ]:
titanic["Embarked"].mode()[0]

In [4]:
titanic["Sex"].mode()[0]

'male'

In [10]:
titanic[['Age', 'Fare']].std()

Age     14.526497
Fare    49.693429
dtype: float64

### Exercise 5 - aggregating statistics grouped by category
---

Refer again to the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)   
looking particularly at the section on Aggregating statistics grouped by category.

1.  What is the mean age for male versus female Titanic passengers?
2.  What is the mean ticket fare price for each of the sex and cabin class combinations?
3.  What is the mean ticket fare price for passengers who embarked at each port?
4.  Which passenger class had the highest number of survivors (for now, just show the statistics - it may not be meaningful yet)?

**Test output**:  
1.  female 27.915709 male 30.726645
2.  
```
female  1         106.125798
            2          21.970121
            3          16.118810
male    1          67.226127
            2          19.741782
            3          12.661633
```
3.  
```
C    59.954144
Q    13.276030
S    27.079812
```
4. 

```
Survived  Pclass

0         1          80
          2          97
          3         372
1         1         136
          2          87
          3         119
```






### Exercise 6 - an aggregation of different statistics
---

Use the function titanic.agg() as shown in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)  

1.  Display

```
     {
         "Age": ["min", "max", "median", "skew"],
         "Fare": ["min", "max", "median", "mean"]
     }
```
2.  Display:  
min, max and mean for Age  
min, max and standard deviation for Fare  
count for Cabin

**Test output**:   
1.  	
```
                  Age	      Fare  
max	   80.000000	512.329200  
mean	  NaN	      32.204208  
median	28.000000	14.454200  
min	   0.420000	 0.000000  
skew	  0.389108	 NaN
```

2.   
```
	        Age	    Fare	   Cabin
count  NaN	    NaN	    204.0
max	80.000000  512.329200 NaN
mean   29.699118  NaN        NaN
min	0.420000   0.000000   NaN
std	NaN        49.693429  NaN
```




### Exercise 7 - count by category
---

Read the section Count number of records by category in the tutorial  
[How to calculate summary statistics?](https://pandas.pydata.org/docs/getting_started/intro_tutorials/06_calculate_statistics.html#)

1. Display the number of passengers of each gender who had a ticket
2. Display the number of passengers who embarked at each port and had a ticket
3. Calculate the percentage of PassengerIds who survived the sinking of the Titanic (*Hint:  try getting the PassengerIds with a count for survived or not.  Store this value in a new variable, which will contain a list/array.  The second item in this list will be the number who survived.  You can use this number and the count of PassengerIds to calculate the percentage*)

**Test output**:  
1.  female 314, male 577
2.  C 168, Q 77, S 644
3.  38.38383838383838



### Exercise 8 - summary happiness statistics
---

Open the data set here: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2019.xlsx?raw=true

It contains data on people's perception of happiness levels in a number of countries across the world.

1.  Display the number of records in the set  
2.  Display the description of the numerical data  
3.  Display the highest GDP and life expectancy  
4.  Display the mean, max and min for Freedom,  mean, max, min and skew for Generosity and mean, min, max and std for GDP  

**Test output**:  
1.  156
2.  Table showing count, mean, std, min, 25%, 50%, 75%, max for 8 columns
3.  GDP 0.905147, life expectancy 0.725244  
4.  


```
	   Freedom to make life choices	Generosity	GDP per capita
max	 0.631000	                   0.566000	  1.684000
mean	0.392571	                   0.184846	  0.905147
min	 0.000000	                   0.000000  	0.000000
skew	NaN	                        0.745942	  NaN
std	 NaN	                        NaN	       0.398389
```




### Exercise 9 - migration data
---

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true  Open the sheet named *Country Migration*

1.  Describe the dataset  
2.  Show summary information
3.  Display the mean net per 10K migration in each of the years 2015 to 2019
4.  Display the mean, max and min migration for the year 2019 for each of the regions (*base_country_wb_region*)
5.  Display the median net migration for the years 2015 and 2019 for the base countries by income level
6.  Display the number of target countries in each income level
7.  Display the mean net migration for all five years, for each income level

**Test output**:  
1  count, mean, std, min, 25%, 50%, 75%, max for 9 columns

2  shows 16 columns with non-null count of 4148 in each column

3  
```
net_per_10K_2015    0.461757
net_per_10K_2016    0.150248
net_per_10K_2017   -0.080272
net_per_10K_2018   -0.040591
net_per_10K_2019   -0.022743
dtype: float64
```

4  
```
	net_per_10K_2019
mean	max	min
base_country_wb_region			
East Asia & Pacific	0.198827	21.57	-9.88
Europe & Central Asia	0.208974	87.71	-21.34
Latin America & Caribbean	-0.904602	21.15	-31.75
Middle East & North Africa	-0.107655	55.60	-50.33
North America	0.239246	23.20	-0.29
South Asia	-0.514577	13.72	-24.89
Sub-Saharan Africa	-0.279729	37.11	-21.54
```

5  
```
	net_per_10K_2015	net_per_10K_2019
base_country_wb_income		
High Income	0.02	0.04
Low Income	0.42	-0.05
Lower Middle Income	-0.02	-0.07
Upper Middle Income	-0.03	-0.08
```

6  
```
base_country_wb_income
High Income            2415
Low Income              185
Lower Middle Income     653
Upper Middle Income     895
Name: target_country_name, dtype: int64 
```

7  
```
net_per_10K_2015	net_per_10K_2016	net_per_10K_2017	net_per_10K_2018	net_per_10K_2019
base_country_wb_income					
High Income	0.505482	0.391379	0.314178	0.379201	0.401470
Low Income	1.876432	0.798270	-0.684865	-0.677784	-0.681459
Lower Middle Income	0.591654	-0.029893	-0.519433	-0.527136	-0.476616
Upper Middle Income	-0.043419	-0.502916	-0.699240	-0.686626	-0.700101
```






In [ ]:
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true"
migration = pd.read_excel(url,sheet_name="Country Migration")
migration.groupby('base_country_wb_income')['net_per_10K_2015','net_per_10K_2016','net_per_10K_2017','net_per_10K_2018','net_per_10K_2019'].mean()


### Exercise 10 - calculating range over a grouped series

Open the dataset at this url: https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true Open the sheet named *Skill Migration*

1.  Display the max for each skill group category of net migration for the year 2017
2.  Assign the max for each skill group category of net migration for the year 2017 to a variable called **max_skill_migration** and print `max_skill_migration`
3.  Create a second variable called **min_skill_migration** and assign to it the min for each skill group category of net migration for the year 2017, print `min_skill_migration`

4.  You now have two series `max_skill_migration` and `min_skill_migration` each of which is a numpy array.  You can perfom calculations on these two series in the same way as you would individual data items.

So, you can calculate the range for each skill category by subtracting the `min_skill_migration` from `max_skill_migration` to get a new series **skill_migration_range**

skill_migration_range = max_skill_migration - min_skill_migration

Try it out.

5.  Now calculate the range for the year 2019
6.  Now calculate the range for countries grouped by income level for the year 2015

Test output:  
1 and 2  
```
skill_group_category
Business Skills                1048.20
Disruptive Tech Skills         1478.56
Soft Skills                    1572.35
Specialized Industry Skills    1906.14
Tech Skills                    1336.78
Name: net_per_10K_2017, dtype: float64
```

3    
```
skill_group_category
Business Skills               -3471.35
Disruptive Tech Skills        -2646.19
Soft Skills                   -2542.23
Specialized Industry Skills   -6604.67
Tech Skills                   -6060.98
Name: net_per_10K_2017, dtype: float64
```

4  
```
skill_group_category
Business Skills                4519.55
Disruptive Tech Skills         4124.75
Soft Skills                    4114.58
Specialized Industry Skills    8510.81
Tech Skills                    7397.76
Name: net_per_10K_2017, dtype: float64
```

5  
```
skill_group_category
Business Skills                4543.96
Disruptive Tech Skills         3651.81
Soft Skills                    5528.47
Specialized Industry Skills    4036.44
Tech Skills                    3424.45
Name: net_per_10K_2019, dtype: float64
```

6  
```
wb_income
High income            4246.50
Low income             4556.42
Lower middle income    2148.36
Upper middle income    4045.43
Name: net_per_10K_2015, dtype: float64
```



